### PR TITLE
feat(test): add adversarial test suite and fix discovered bugs

### PR DESCRIPTION
**Pull request by jamesbrink**
## Summary

- Add ~860 tests across the workspace targeting coverage gaps, edge cases, and adversarial inputs
- Introduce `/blackbox-test` skill for generating black-box integration tests from public API signatures
- Fix 4 real bugs discovered by the new tests
- Coverage: **80.3% -> 83.7%** lines across claudette + claudette-server

## Bugs fixed

- **`build_persistent_args` effort validation**: was passing arbitrary strings (empty, "auto", "garbage") to CLI unvalidated; now matches `build_claude_args` validation
- **Auth session limit**: `pair()` allowed unbounded session accumulation; now caps at `MAX_SESSIONS` (64) with oldest-first eviction via `while` loop (handles legacy oversized configs)
- **TLS cert parse errors**: `filter_map(|r| r.ok())` silently dropped corrupt cert parse errors; now propagates via `collect::<Result>()`
- **Usage test**: `stale_cache_triggers_refetch_but_falls_back` was broken because mock cache triggered HTTP token refresh

## Test suites added

| Area | Tests | Coverage |
|------|-------|----------|
| claudette-server auth | 38 | 0% -> 99% |
| claudette-server handler | 20 | 0% -> 21% |
| claudette-server TLS | 13 | 0% -> 100% |
| auth date algorithms | 11 | inline |
| agent persistent args | 12 | validation gap |
| MCP detection | 13 | 88% -> 93% |
| git error paths | 10 | 92% -> 97% |
| usage.rs | 2 new + 1 fix | stale cache |
| Black-box integration | ~700 | agent, db, diff, mcp, model, names, config, etc. |

## Test plan

- [x] `cargo fmt --all --check` -- clean
- [x] `cargo clippy --workspace --all-targets -- -D warnings` -- zero warnings
- [x] `cargo test --workspace --all-features` -- 860 tests, 0 failures
- [x] Codex peer review -- 3 findings addressed

### CODE REVIEW

File: tests/test_diff.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    // The behavior depends on the parser -- just verify no panic
+    let _ = result;
+}
+
+/// Completely garbage input that looks nothing like a diff.
+#[test]
+fn test_diff_parse_garbage_input() {
+    let result = parse_unified_diff("this is not a diff at all\nrandom text\n42\n", "garbage.rs");
+    // Should return empty or ignore non-diff content
+    assert!(result.hunks.is_empty() || !result.hunks.is_empty()); // just no panic
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply tests/test_diff.rs]: Fixed. Replaced the tautology with concrete assertions: `hunks.is_empty()`, `path == "garbage.rs"`, `!is_binary`.

File: tests/test_agent.rs [hunk 2]
```diff
@@ -0,0 +1,1740 @@
...
+}
+
+// ─── sanitize_branch_name tests ─────────────────────────────────────
+
+/// Empty input should return an empty string or some default.
+#[test]
+fn test_agent_sanitize_branch_name_empty() {
+    let result = sanitize_branch_name("", 50);
+    // Empty input after sanitization -- should be empty
+    assert!(result.is_empty() || !result.is_empty()); // just no panic
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply tests/test_agent.rs]: Fixed. Now asserts `result == ""` directly.

File: tests/suite/test_names.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    // The display format might capitalize or use different separators,
+    // but should contain both words in some form
+    let display_lower = name.display.to_lowercase();
+    assert!(
+        display_lower.contains(&name.adjective) || display_lower.contains(&name.plant),
+        "Display '{}' should contain adjective '{}' or plant '{}'",
+        name.display,
+        name.adjective,
+        name.plant,
+    );
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply tests/suite/test_names.rs:182]: Good catch. Changed `||` to `&&` and updated the message to say "and" instead of "or".

File: src/agent.rs [hunk 4]
```diff
@@ -1177,6 +1237,96 @@ pub async fn generate_branch_name(
...
     use super::*;
+    use std::os::unix::fs::PermissionsExt;
+    use tokio::time::{Duration, timeout};
+
+    fn make_fake_claude_script(dir: &Path) -> std::path::PathBuf {
+        let path = dir.join("claude");
+        std::fs::write(
+            &path,
+            r#"#!/bin/sh
+set -eu
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/agent.rs:1245]: This is pre-existing code (not introduced by this PR) and the project only targets macOS + Linux per CLAUDE.md. Windows is not a supported platform.

File: src/mcp.rs [hunk 5]
```diff
@@ -289,7 +289,7 @@ fn detect_user_global_mcps() -> Option<Vec<McpServer>> {
 }
 
 /// Parse committed `.mcp.json` at the repository root.
-fn detect_project_mcp_json(repo_path: &Path) -> Option<Vec<McpServer>> {
+pub fn detect_project_mcp_json(repo_path: &Path) -> Option<Vec<McpServer>> {
     let config_path = repo_path.join(".mcp.json");
     let content = std::fs::read_to_string(&config_path).ok()?;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/mcp.rs:296]: Valid concern. Added `#[doc(hidden)]` to both functions to signal they are pub for integration tests only, not part of the stable API.

File: src-server/src/auth.rs [hunk 6]
```diff
@@ -89,6 +92,12 @@ impl ServerConfig {
             created_at: now.clone(),
             last_seen: now,
         });
+        // Truncate oldest sessions to stay within the limit.
+        // Handles both normal growth (one over) and legacy configs
+        // that already exceed MAX_SESSIONS.
+        while self.sessions.len() > MAX_SESSIONS {
+            self.sessions.remove(0);
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-server/src/auth.rs]: Fixed. Replaced `while { remove(0) }` with `drain(0..excess)` for O(n) eviction.

File: src-server/tests/test_handler.rs [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    let cert_pem = cert.pem();
+    let key_pem = key_pair.serialize_pem();
+    std::fs::write(cert_dir.join("cert.pem"), &cert_pem).unwrap();
+    std::fs::write(cert_dir.join("key.pem"), &key_pem).unwrap();
+
+    // Server config.
+    let certs: Vec<rustls::pki_types::CertificateDer<'static>> =
+        rustls_pemfile::certs(&mut cert_pem.as_bytes())
+            .filter_map(|r| r.ok())
+            .collect();
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-server/tests/test_handler.rs]: Fixed. Now uses `collect::<Result<Vec<_>, _>>().expect(...)` matching the production code pattern.

File: src/mcp.rs [hunk 8]
```diff
@@ -435,7 +437,7 @@ fn detect_user_project_mcps(repo_path: &Path) -> Option<Vec<McpServer>> {
 
 /// Parse MCPs from `{repo}/.claude.json`, but only if it's explicitly
 /// gitignored.
-fn detect_repo_local_mcps(repo_path: &Path) -> Option<Vec<McpServer>> {
+pub fn detect_repo_local_mcps(repo_path: &Path) -> Option<Vec<McpServer>> {
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/mcp.rs:440]: `detect_repo_local_mcps` was not actually made pub — only `normalize_mcp_config` and `detect_project_mcp_json` were, and both now have `#[doc(hidden)]`. No change needed.

File: src-server/src/handler.rs [hunk 9]
```diff
@@ -817,3 +817,112 @@ async fn handle_set_app_setting(
...
+        }
+    }
+
+    #[test]
+    fn test_now_iso_reasonable_year() {
+        let result = now_iso();
+        let year: u32 = result[..4].parse().expect("first 4 chars should be a year");
+        assert!(
+            (2024..=2100).contains(&year),
+            "now_iso() year {year} is outside 2024..2100 range"
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-server/src/handler.rs]: Fixed. Replaced hardcoded 2024..=2100 range with `year >= 1970` check.

File: src-server/src/auth.rs [hunk 10]
```diff
@@ -134,3 +144,132 @@ fn epoch_days_to_date(days: u64) -> (u64, u64, u64) {
...
+    }
+
+    #[test]
+    fn test_now_iso_reasonable_year() {
+        let ts = now_iso();
+        let year: u64 = ts[..4].parse().expect("first 4 chars should be a year");
+        assert!(
+            (2024..=2100).contains(&year),
+            "year {year} is outside the expected range 2024..=2100"
+        );
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-server/src/auth.rs:273]: Fixed. Replaced hardcoded 2024..=2100 range with `year >= 1970` check.

File: tests/test_names.rs [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    let _ = name;
+}
+
+/// Add a wildcard easter egg.
+#[test]
+fn test_names_easter_egg_wildcard() {
+    let mut ng = NameGenerator::new();
+    let plant = PLANTS[0];
+    ng.add_wildcard_egg("*", plant, EasterEgg::Message("any adj!".to_string()));
+    let _ = ng.generate();
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply tests/test_names.rs]: Fixed. Now uses `generate_from_seed(0)` which deterministically selects `PLANTS[0]`, and asserts `easter_egg.is_some()`.

File: src-tauri/src/usage.rs [hunk 12]
```diff
@@ -814,4 +816,46 @@ mod tests {
...
+        let original_fetched_at = now_millis() - USAGE_CACHE_TTL_MS - 1;
+        let cache = make_cache_with_usage(original_fetched_at);
+
+        // The stale fallback should succeed.
+        let result = rt.block_on(get_usage(&cache)).unwrap();
+        // The returned data should have the original fetched_at from the
+        // cached ClaudeCodeUsage, not a new timestamp.
+        assert_eq!(
+            result.fetched_at, 0,
+            "Stale fallback returns the original cached data"
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-tauri/src/usage.rs:858]: Fixed. Changed `fetched_at` from `0` to `1_700_000_000_000` (a realistic timestamp) and updated the assertion to match.

File: src-server/src/tls.rs [hunk 13]
```diff
@@ -19,8 +19,10 @@ pub fn load_or_generate_tls(
     let key_pem = std::fs::read(&key_path)?;
 
     let certs: Vec<CertificateDer<'static>> = rustls_pemfile::certs(&mut &cert_pem[..])
-        .filter_map(|r| r.ok())
-        .collect();
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|e| -> Box<dyn std::error::Error> {
+            format!("Failed to parse cert.pem: {e}").into()
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src-server/src/tls.rs:24]: Invalid. `String` implements `Into<Box<dyn Error>>` via the blanket impl `impl<E: Error> From<E> for Box<dyn Error>` combined with `impl From<String> for Box<dyn Error>` in std. This compiles and works correctly.

File: tests/test_permissions.rs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    // If case-sensitive, these will differ
+    let _ = (full, upper);
+}
+
+/// "default" or "standard" level if it exists.
+#[test]
+fn test_permissions_default_level() {
+    let tools = tools_for_level("default");
+    // Should not panic; check it returns something reasonable
+    let _ = tools;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply tests/test_permissions.rs]: Fixed. All permission tests now have concrete assertions: `assert_eq!` against known return values, `assert_ne!` for case sensitivity, specific tool membership checks.

File: tests/test_names.rs [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+#[test]
+fn test_names_easter_egg_exact() {
+    let mut ng = NameGenerator::new();
+    let adj = ADJECTIVES[0];
+    let plant = PLANTS[0];
+    ng.add_egg(adj, plant, EasterEgg::Message("found it!".to_string()));
+
+    let name = ng.generate_from_seed(0);
+    // The seed may or may not hit this combo, but this tests that add_egg doesn't panic
+    let _ = name;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply tests/test_names.rs]: Fixed. Now uses `generate_from_seed(0)` and asserts `easter_egg.is_some()`.
